crates/typst-syntax/src/lexer.rs

```diff
@@ -561,11 +564,48 @@ impl Lexer<'_> {
             // Identifiers.
             c if is_math_id_start(c) && self.s.at(is_math_id_continue) => {
                 self.s.eat_while(is_math_id_continue);
-                SyntaxKind::MathIdent
+                return self.math_ident_or_field(start);
             }
 
             // Other math atoms.
             _ => self.math_text(start, c),
+        };
+        (kind, None)
+    }
+
+    /// Parse a single MathIdent or an entire FieldAccess SyntaxNode.
+    fn math_ident_or_field(&mut self, start: usize) -> (SyntaxKind, Option<SyntaxNode>) {
+        let initial_ident = self.s.from(start);
+        let Some(mut next_ident) = self.maybe_dot_ident() else {
+            return (SyntaxKind::MathIdent, None);
+        };
+        let mut node = SyntaxNode::leaf(SyntaxKind::MathIdent, initial_ident);
+        loop {
+            let field_children = vec![
+                node,
+                SyntaxNode::leaf(SyntaxKind::Dot, '.'),
+                SyntaxNode::leaf(SyntaxKind::Ident, next_ident),
+            ];
+            node = SyntaxNode::inner(SyntaxKind::FieldAccess, field_children);
+            match self.maybe_dot_ident() {
+                None => break,
+                Some(ident) => next_ident = ident,
+            }
+        }
+        (SyntaxKind::FieldAccess, Some(node))
```

Comment:
That's a clever rewrite with the while. I like it.

However, I would like to keep `maybe_dot_ident` as a separate function. Mostly just to keep the scanner logic separate properly from the node-building logic, and because it feels like a proper function.